lib/logstash/filters/parser.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+require 'digest'
+
+# Define a Parser class for processing tokens and generating templates.
+class Parser
+  def initialize(tokens_list, single_dict, double_dict, tri_dict, threshold)
+    @tokens_list = tokens_list
+    @single_dict = single_dict
+    @double_dict = double_dict
+    @tri_dict = tri_dict
```

Comment:
Why do we need to initialize this? We should grab it from `gram_dict` whenever we need. Let's pass in `gram_dict` as a parameter